rpcs3/Emu/Io/Dimensions.cpp

```diff
@@ -0,0 +1,504 @@
+	m_figure_added_removed_responses.pop();
+	return response;
+}
+
+usb_device_dimensions::usb_device_dimensions(const std::array<u8, 7>& location)
+	: usb_device_emulated(location)
+{
+	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x200, 0x0, 0x0, 0x0, 0x20, 0x0E6F, 0x0241, 0x200, 0x1, 0x2, 0x3, 0x1});
+	auto& config0 = device.add_node(UsbDescriptorNode(USB_DESCRIPTOR_CONFIG, UsbDeviceConfiguration{0x29, 0x1, 0x1, 0x0, 0x80, 0xFA}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_INTERFACE, UsbDeviceInterface{0x0, 0x0, 0x2, 0x3, 0x0, 0x0, 0x0}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_HID, UsbDeviceHID{0x0111, 0x00, 0x01, 0x22, 0x001d}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x81, 0x03, 0x20, 0x1}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x01, 0x03, 0x20, 0x1}));
+}
+
+usb_device_dimensions::~usb_device_dimensions()
+{
+}
+
+void usb_device_dimensions::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue, u16 wIndex, u16 wLength, u32 buf_size, u8* buf, UsbTransfer* transfer)
+{
+	usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+}
+
+void usb_device_dimensions::interrupt_transfer(u32 buf_size, u8* buf, u32 endpoint, UsbTransfer* transfer)
+{
+	ensure(buf_size == 0x20);
+
+	if (endpoint == 0x81)
+	{
+		// Read Endpoint, it is required to wait until there is a response to reply with
+		// otherwise the game believes that the portal is disconnected. Loop through queries
+		// in a seperate thread, then use fake transfer callback method to push the response
+		// to the list of other fake response
+		std::thread read_message([this, buf, transfer, buf_size]()
+			{
+				bool responded = false;
+				do
+				{
+					std::unique_lock lock(query_mutex);
+					if (g_dimensionstoypad.has_figure_been_added_removed())
+					{
+						memcpy(buf, g_dimensionstoypad.pop_added_removed_response().data(), 0x20);
+						responded = true;
+					}
+					else if (!m_queries.empty())
+					{
+						memcpy(buf, m_queries.front().data(), 0x20);
+						m_queries.pop();
+						responded = true;
+					}
+					lock.unlock();
+					std::this_thread::sleep_for(100ms);
+				} while (!responded);
+
+				fake_callback_transfer(transfer, buf_size);
+			});
+		read_message.detach();
+	}
+	else if (endpoint == 0x01)
+	{
+		dimensions_log.error("Request: %s", fmt::buf_to_hexstring(buf, buf_size));
+		// Write endpoint, similar structure of request to the Infinity Base with a command for byte 3,
+		// sequence for byte 4, the payload after that, then a checksum for the final byte.
+		transfer->fake = true;
+		transfer->expected_count = buf_size;
+		transfer->expected_result = HC_CC_NOERR;
+
+		const u8 command = buf[2];
+		const u8 sequence = buf[3];
+
+		transfer->expected_time = get_timestamp() + 100;
+		std::array<u8, 32> q_result{};
+
+		switch (command)
+		{
+		// Wake
+		case 0xB0:
```

Comment:
I think it would be easier to read if all of these comments were to the right of the values
```suggestion
		case 0xB0: // Wake
```